docs/proposals/horizontal-pod-autoscaler.md

```diff
@@ -0,0 +1,272 @@
+	api.ObjectMeta
+
+	// Spec defines the behaviour of autoscaler.
+	Spec HorizontalPodAutoscalerSpec
+
+	// Status represents the current information about the autoscaler.
+	Status HorizontalPodAutoscalerStatus
+}
+
+// HorizontalPodAutoscalerSpec is the specification of a horizontal pod autoscaler.
+type HorizontalPodAutoscalerSpec struct {
+	// ScaleRef is a reference to Scale subresource. HorizontalPodAutoscaler will learn the current
+	// resource consumption from its status, and will set the desired number of pods by modyfying its spec.
+	ScaleRef *SubresourceReference
+	// MinCount is the lower limit for the number of pods that can be set by the autoscaler.
+	MinCount int
+	// MaxCount is the upper limit for the number of pods that can be set by the autoscaler.
+	// It cannot be smaller than MinCount.
+	MaxCount int
+	// Target is the target average consumption of the given resource that the autoscaler will try
+	// to maintain by adjusting the desired number of pods.
+	// Currently two types of resources are supported: "cpu" and "memory".
+	Target ResourceConsumption
+}
+
+// HorizontalPodAutoscalerStatus contains the current status of a horizontal pod autoscaler
+type HorizontalPodAutoscalerStatus struct {
+	// CurrentReplicas is the number of replicas of pods managed by this autoscaler.
+	CurrentReplicas int
+
+	// DesiredReplicas is the desired number of replicas of pods managed by this autoscaler.
+	// The number may be different because pod downscaling is someteimes delayed to keep the number
+	// of pods stable.
+	DesiredReplicas int
+
+	// CurrentConsumption is the current average consumption of the given resource that the autoscaler will
+	// try to maintain by adjusting the desired number of pods.
+	// Two types of resources are supported: "cpu" and "memory".
+	CurrentConsumption ResourceConsumption
+
+	// LastScaleTimestamp is the last time the HorizontalPodAutoscaler scaled the number of pods.
+	// This is used by the autoscaler to controll how often the number of pods is changed.
+	LastScaleTimestamp *util.Time
+}
+
+// ResourceConsumption is an object for specifying average resource consumption of a particular resource.
+type ResourceConsumption struct {
+	Resource api.ResourceName
+	Quantity resource.Quantity
+}
+```
+
+```Scale``` will be a reference to the Scale subresource.
+```MinCount```, ```MaxCount``` and ```Target``` will define autoscaler configuration.
+We will also introduce HorizontalPodAutoscalerList object to enable listing all autoscalers in the cluster:
+
+```go
+// HorizontalPodAutoscaler is a collection of pod autoscalers.
+type HorizontalPodAutoscalerList struct {
+	api.TypeMeta
+	api.ListMeta
+
+	Items []HorizontalPodAutoscaler
+}
+```
+
+## Autoscaling Algorithm
+
+The autoscaler will be implemented as a control loop.
+It will periodically (e.g.: every 1 minute) query pods described by ```Status.PodSelector``` of Scale subresource,
+and check their average CPU or memory usage from the last 1 minute
```

Comment:
Do we know if it makes sense to use memory as a signal for scaling? What would cause an application to consume more memory under higher load? More simultaneous connections thus more total connection state (including things like caches)?  I assume horizontal pod autoscaling primarily targets stateless front-ends, and stateful applications that can dynamically re-shard. I don't know enough about these apps to know whether memory is a useful metric to use to trigger scaling.